src/shared/components/query/GenesetsJsTree.tsx

```diff
@@ -0,0 +1,216 @@
+import React from 'react';
+import $ from 'jquery';
+import {getHierarchyData} from "shared/lib/StoreUtils";
+//Import jstree.min and style for jstree to work
+import 'jstree/dist/jstree.min'; // tslint:disable-line
+import 'shared/components/query/styles/jstree/style.css'; // tslint:disable-line
+import * as _ from "lodash";
+import { remoteData } from "shared/api/remoteData";
+import CBioPortalAPIInternal, { GenesetHierarchyInfo } from "shared/api/generated/CBioPortalAPIInternal";
+import { observer } from "mobx-react";
+import { observable, ObservableMap } from "mobx";
+import LoadingIndicator from "shared/components/loadingIndicator/LoadingIndicator";
+
+export interface GenesetsJsTreeProps
+{
+    initialSelection: string[];
+    scoreThreshold: string;
+    pvalueThreshold: string;
+    percentile: string;
+    gsvaProfile: string;
+    sampleListId: string|undefined;
+    searchValue: string;
+    onSelect: (map_geneSet_selected:ObservableMap<boolean>) => void;
+}
+
+type JSNode = {
+    original: {description: string, geneset: boolean};
+};
+
+@observer
+export default class GenesetsJsTree extends React.Component<GenesetsJsTreeProps, {}>
+{
+    tree: Element|null;
+    @observable isLoading = true;
+    promisedTree: Promise<Element>;
+    private readonly map_geneSets_selected = new ObservableMap<boolean>();
+
+    constructor(props: GenesetsJsTreeProps)
+    {
+        super(props);
+        this.map_geneSets_selected.replace(props.initialSelection.map(geneSet => [geneSet, true]));
+    }
+    
+    componentDidMount(){
+        this.promisedTree = this.initTree(this.tree as Element); //React only sets the div to null when the component unmounts
+        this.promisedTree.then(tree => (this.isLoading = false, tree));
+    }
+    
+    componentDidUpdate(prevProps: GenesetsJsTreeProps){
```

Comment:
this comparison may already be accomplished by the built in implementation of shouldComponentUpdate provided by @observer decorator (react-mobx).   If so, then you might not need this logic.  I.e. component will not update unless properties are actually changed.

Reply:
if logic is more complex than than, i understand.

Reply:
According to the [React documentation](https://reactjs.org/docs/react-component.html#componentdidupdate), `componentDidUpdate()` gets called whenever *any* one of the component's properties changes, and care should be taken not to waste resources on network requests if the changed properties are irrelevant to the network request (such as in this case, `searchValue`). That's the reason `componentDidUpdate()` accepts the previous props and state as arguments.

Reply:
if you read comment again, i'm point out that this optimization may already be accomplished by @observer, which implements shouldComponent update.  If it's true that there are properties of this component which can change without necessitating another network request, then I understand this more granular logic necessary.

Reply:
@alisman But I need to call a function or another depending on which props do update. Can `shouldComponentUpdate` do that? I couldn't find any example.

Reply:
No, if it's the case you need to call a specific function then this is the way to do it.